File: examples/memos-sample/validation/memos-flow.js
'use strict';

// Auth and business actions use public Bridge calls. The SQLite oracle runs in
// a separate process after this Script; it never supplies UI success signals.
const fs = require('node:fs');
const elementKeys = ['elementId', 'tag', 'id', 'name', 'type', 'role', 'ariaLabel', 'placeholder', 'href', 'text'];

module.exports.main = async function main(ctx) {
  const { marker, content, editedContent, credentialsPath, negativeOnly = false } = ctx.inputs;
  const verdicts = [];
  async function call(command, args = {}) {
    const result = await ctx.call(command, args);
    if (!result.ok || result.ambiguous === true) throw new Error(`${command}:${result.error || 'ambiguous'}`);
    return result;
  }
  const read = selector => call('web-dom', selector ? { selector } : {});
  const controls = observation => observation.result.dom.controls.filter(node => node.visible);
  function one(observation, predicate, label) {
    const nodes = controls(observation).filter(predicate);
    if (nodes.length !== 1) throw new Error(`${label}:expected one observed control, received ${nodes.length}`);
    return nodes[0];
  }
  async function act(observation, node, command, args = {}) {
    return call(command, { selector: { elementId: node.elementId }, expectedTarget: {
      pageRef: observation.result.pageRef,
      element: Object.fromEntries(elementKeys.map(key => [key, node[key]])),
    }, ...args });
  }
  async function check(name, observation, condition) {
    const verdict = await ctx.assert({ name, condition, requiredEvidence: ['tree'], evidence: observation.evidence });
    verdicts.push(verdict);
    if (verdict.verdict !== 'passed') throw new Error(`${name}:${verdict.verdict}:${verdict.reason || ''}`);
  }
  async function until(name, selector, predicate) {
    const deadline = Date.now() + 15000;
    do {
      const observation = await read(selector);
      if (predicate(observation)) { await check(name, observation, true); return observation; }
      await new Promise(resolve => setTimeout(resolve, 200));
    } while (Date.now() < deadline);
    throw new Error(`${name}:observed state deadline`);
  }
  async function button(text) {
    const observation = await until(`${text} is ready`, undefined, value => {
      const nodes = controls(value).filter(node => node.text === text);
      return nodes.length === 1 && nodes[0].interaction.status === 'ready';
    });
    return act(observation, one(observation, node => node.text === text, text), 'web-click');
  }

  let observation = await read();
  if (new URL(observation.result.dom.url).pathname === '/auth') {
    const credentials = JSON.parse(fs.readFileSync(credentialsPath, 'utf8'));
    await act(observation, one(observation, node => node.id === 'signin-username', 'username'), 'web-input', { value: credentials.username });
    observation = await read();
    await act(observation, one(observation, node => node.id === 'signin-password', 'password'), 'web-input', { value: credentials.password });
    await button('登录');
    observation = await until('authenticated home', undefined, value => new URL(value.result.dom.url).pathname === '/');
  }
  if (negativeOnly) {
    await check('wrong memo content must fail', observation, observation.result.dom.bodyText.includes(marker));
    return { verdicts };
  }
  observation = await until('home editor is ready', undefined, value => controls(value).some(node =>
    node.role === 'textbox' && node.editable && node.interaction.status === 'ready'));
  await check('fresh marker is absent before create', observation, !observation.result.dom.bodyText.includes(marker));
  await act(observation, one(observation, node => node.role === 'textbox' && node.editable, 'home editor'), 'web-input', { value: content });
  await button('Save');
  await until('created memo is rendered as an article', 'article', value => controls(value).some(node => node.text.includes(marker)));

  observation = await read();
  await act(observation, one(observation, node => node.tag === 'input' && node.placeholder === 'Search memos...', 'search'), 'web-input', { value: marker });
  observation = await read();
  await act(observation, one(observation, node => node.tag === 'input' && node.placeholder === 'Search memos...', 'search'), 'web-key', { key: 'Enter' });
  await until('search isolates the created memo', 'article', value => controls(value).length === 1 && controls(value)[0].text.includes(marker));

  async function openMenu() {
    const article = await read('article button');
    return act(article, one(article, node => node.tag === 'button', 'memo menu'), 'web-click');
  }
  await openMenu(); await button('Edit');
  observation = await read();
  const editor = one(observation, node => node.role === 'textbox' && node.editable && node.text.includes(marker), 'existing memo editor');
  await check('existing editor contains original content', observation, editor.text.includes(marker) && editor.text.includes('牛奶 2 瓶'));
  await act(observation, editor, 'web-input', { value: editedContent });
  observation = await read();
  await act(observation, one(observation, node => node.tag === 'button' && node.text === 'Save' && !node.disabled,
    'enabled edit Save'), 'web-click');
  await until('edited memo is rendered', 'article', value => controls(value).length === 1 && controls(value)[0].text.includes('Script edit verified'));

  await openMenu(); await button('Delete');
  observation = await read();
  await check('delete requires confirmation', observation,
    controls(observation).some(node => node.text === 'Cancel') && controls(observation).some(node => node.text === 'Delete'));
  await button('Cancel');
  observation = await until('cancelled delete retains the memo', undefined, value =>
    value.result.dom.bodyText.includes(marker) && !controls(value).some(node => node.tag === 'button' && node.text === 'Cancel'));
  await check('final edited business content is visible', observation, observation.result.dom.bodyText.includes('Script edit verified'));
  return { marker, verdicts, assertionCount: verdicts.length };
};

File: examples/memos-sample/validation/memos-flow.py
"""Python equivalent of memos-flow.js, using public Bridge calls only.

The independent SQLite oracle runs after this Script and never supplies UI
success signals. Inputs use a unique marker and the same full-content contract.
"""
import json
from pathlib import Path
import time
from urllib.parse import urlparse


ELEMENT_KEYS = ('elementId', 'tag', 'id', 'name', 'type', 'role', 'ariaLabel',
                'placeholder', 'href', 'text')


def main(ctx):
    marker = ctx.inputs['marker']
    verdicts = []

    def call(command, args=None):
        result = ctx.call(command, {} if args is None else args)
        if result['ok'] is not True or result.get('ambiguous') is True:
            raise RuntimeError(f"{command}:{result.get('error')}")
        return result

    def read(selector=None):
        return call('web-dom', {} if selector is None else {'selector': selector})

    def controls(observation):
        return [node for node in observation['result']['dom']['controls'] if node['visible']]

    def one(observation, predicate, label):
        nodes = [node for node in controls(observation) if predicate(node)]
        if len(nodes) != 1:
            raise RuntimeError(f'{label}:expected one observed control, received {len(nodes)}')
        return nodes[0]

    def act(observation, node, command, args=None):
        return call(command, {
            'selector': {'elementId': node['elementId']},
            'expectedTarget': {'pageRef': observation['result']['pageRef'],
                               'element': {key: node[key] for key in ELEMENT_KEYS}},
            **({} if args is None else args),
        })

    def check(name, observation, condition):
        verdict = ctx.assert_({'name': name, 'condition': condition,
                               'requiredEvidence': ['tree'], 'evidence': observation['evidence']})
        verdicts.append(verdict)
        if verdict['verdict'] != 'passed':
            raise RuntimeError(f"{name}:{verdict['verdict']}:{verdict.get('reason')}")

    def until(name, selector, predicate):
        deadline = time.monotonic() + 15
        while True:
            observation = read(selector)
            if predicate(observation):
                check(name, observation, True)
                return observation
            if time.monotonic() >= deadline:
                raise RuntimeError(f'{name}:observed state deadline')
            time.sleep(0.2)

    def button(text):
        def ready(value):
            nodes = [node for node in controls(value) if node['text'] == text]
            return len(nodes) == 1 and nodes[0]['interaction']['status'] == 'ready'
        observation = until(f'{text} is ready', None, ready)
        return act(observation, one(observation, lambda node: node['text'] == text, text), 'web-click')

    observation = read()
    if urlparse(observation['result']['dom']['url']).path == '/auth':
        credentials = json.loads(Path(ctx.inputs['credentialsPath']).read_text())
        act(observation, one(observation, lambda node: node['id'] == 'signin-username', 'username'),
            'web-input', {'value': credentials['username']})
        observation = read()
        act(observation, one(observation, lambda node: node['id'] == 'signin-password', 'password'),
            'web-input', {'value': credentials['password']})
        button('Sign in')
        observation = until('authenticated home', None,
                            lambda value: urlparse(value['result']['dom']['url']).path == '/')
    if ctx.inputs.get('negativeOnly') is True:
        check('wrong memo content must fail', observation, marker in observation['result']['dom']['bodyText'])
        return {'verdicts': verdicts}

    observation = until('home editor is ready', None, lambda value: any(
        node['role'] == 'textbox' and node['editable'] and node['interaction']['status'] == 'ready'
        for node in controls(value)))
    check('fresh marker is absent before create', observation, marker not in observation['result']['dom']['bodyText'])
    act(observation, one(observation, lambda node: node['role'] == 'textbox' and node['editable'], 'home editor'),
        'web-input', {'value': ctx.inputs['content']})
    button('Save')
    until('created memo is rendered as an article', 'article',
          lambda value: any(marker in node['text'] for node in controls(value)))

    observation = read()
    search = lambda node: node['tag'] == 'input' and node['placeholder'] == 'Search memos...'
    act(observation, one(observation, search, 'search'), 'web-input', {'value': marker})
    observation = read()
    act(observation, one(observation, search, 'search'), 'web-key', {'key': 'Enter'})
    until('search isolates the created memo', 'article',
          lambda value: len(controls(value)) == 1 and marker in controls(value)[0]['text'])

    def open_menu():
        article = read('article button')
        return act(article, one(article, lambda node: node['tag'] == 'button', 'memo menu'), 'web-click')

    open_menu()
    button('Edit')
    observation = read()
    editor = one(observation, lambda node: node['role'] == 'textbox' and node['editable'] and marker in node['text'],
                 'existing memo editor')
    check('existing editor contains original content', observation,
          marker in editor['text'] and '牛奶 2 瓶' in editor['text'])
    act(observation, editor, 'web-input', {'value': ctx.inputs['editedContent']})
    observation = read()
    act(observation, one(observation, lambda node: node['tag'] == 'button' and node['text'] == 'Save' and not node['disabled'],
                         'enabled edit Save'), 'web-click')
    until('edited memo is rendered', 'article', lambda value: len(controls(value)) == 1
          and 'Script edit verified' in controls(value)[0]['text'])

    open_menu()
    button('Delete')
    observation = read()
    check('delete requires confirmation', observation,
          any(node['text'] == 'Cancel' for node in controls(observation))
          and any(node['text'] == 'Delete' for node in controls(observation)))
    button('Cancel')
    observation = until('cancelled delete retains the memo', None,
                        lambda value: marker in value['result']['dom']['bodyText'] and not any(
                            node['tag'] == 'button' and node['text'] == 'Cancel' for node in controls(value)))
    check('final edited business content is visible', observation,
          'Script edit verified' in observation['result']['dom']['bodyText'])
    return {'marker': marker, 'verdicts': verdicts, 'assertionCount': len(verdicts)}
